src/coreclr/src/tools/Common/TypeSystem/Common/Utilities/GCPointerMap.Algorithm.cs

```diff
@@ -0,0 +1,124 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Debug = System.Diagnostics.Debug;
+
+namespace Internal.TypeSystem
+{
+    partial struct GCPointerMap
+    {
+        /// <summary>
+        /// Computes the GC pointer map for the instance fields of <paramref name="type"/>.
+        /// </summary>
+        public static GCPointerMap FromInstanceLayout(DefType type)
+        {
+            Debug.Assert(type.ContainsGCPointers);
+
+            GCPointerMapBuilder builder = new GCPointerMapBuilder(type.InstanceByteCount.AsInt, type.Context.Target.PointerSize);
+            FromInstanceLayoutHelper(ref builder, type);
+
+            return builder.ToGCMap();
+        }
+
+        /// <summary>
+        /// Computes the GC pointer map of the GC static region of the type.
+        /// </summary>
+        public static GCPointerMap FromStaticLayout(DefType type)
+        {
+            GCPointerMapBuilder builder = new GCPointerMapBuilder(type.GCStaticFieldSize.AsInt, type.Context.Target.PointerSize);
+
+            foreach (FieldDesc field in type.GetFields())
+            {
+                if (!field.IsStatic || field.HasRva || field.IsLiteral
+                    || field.IsThreadStatic || !field.HasGCStaticBase)
+                    continue;
+
+                TypeDesc fieldType = field.FieldType;
+                if (fieldType.IsGCPointer)
+                {
+                    builder.MarkGCPointer(field.Offset.AsInt);
+                }
+                else
+                {
+                    Debug.Assert(fieldType.IsValueType);
+                    var fieldDefType = (DefType)fieldType;
+                    if (fieldDefType.ContainsGCPointers)
+                    {
+                        GCPointerMapBuilder innerBuilder =
+                            builder.GetInnerBuilder(field.Offset.AsInt, fieldDefType.InstanceByteCount.AsInt);
+                        FromInstanceLayoutHelper(ref innerBuilder, fieldDefType);
+                    }
+                }
+            }
+
+            Debug.Assert(builder.ToGCMap().Size * type.Context.Target.PointerSize >= type.GCStaticFieldSize.AsInt);
+            return builder.ToGCMap();
+        }
+
+        /// <summary>
+        /// Computes the GC pointer map of the thread static region of the type.
+        /// </summary>
+        public static GCPointerMap FromThreadStaticLayout(DefType type)
```

Comment:
Let's delete the threadstatics and GC-statics parts of this file. They'll probably look different for CoreCLR anyway and this file is already partial.

Reply:
Done